File: go/blackjack/strategy/soft.go
package strategy

import (
	"github.com/bnwest/GoBlackjackSimulation/go/blackjack/cards"
)

var softTotalDecision = [22][14]Decision{
	//0   A   2   3   4   5   6   7   8   9  10   J   Q   K
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 1  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 2  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 3  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 4  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 5  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 6  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 7  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 8  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 9  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 10  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 11  x  dealer top card
	{NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO},
	// soft total: 12 (A, A)  x  dealer top card
	{NO,  H,  H,  H,  H,  H, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 13 (A, 2)  x  dealer top card
	{NO,  H,  H,  H,  H, Dh, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 14 (A, 3)  x  dealer top card
	{NO,  H,  H,  H,  H, Dh, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 15 (A, 4)  x  dealer top card
	{NO,  H,  H,  H, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 16 (A, 5)  x  dealer top card
	{NO,  H,  H,  H, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 17 (A, 6)  x  dealer top card
	{NO,  H,  H, Dh, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H},
	// soft total: 18 (A, 7)  x  dealer top card
	{NO,  H, Ds, Ds, Ds, Ds, Ds,  S,  S,  H,  H,  H,  H,  H},
	// soft total: 19 (A, 8)  x  dealer top card
	{NO,  S,  S,  S,  S,  S, Ds,  S,  S,  S,  S,  S,  S,  S},
	// soft total: 20 (A, 9)  x  dealer top card
	{NO,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S},
	// soft total: 21 (A, 10)  x  dealer top card
	{NO,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S},
	//0   A   2   3   4   5   6   7   8   9  10   J   Q   K
}

func createSoftTotalDecisions() [22]map[cards.CardRank]Decision {
	// Convert 2D array of Decisions into a list of maps:
	// decisions = {
	//     0: {cards.ACE: NO, ..,}, ...
	// }
	decisions := [22]map[cards.CardRank]Decision{}

	// initialize iner map
	for i := 0; i < 22; i++ {
		decisions[i] = make(map[cards.CardRank]Decision)
	}

	for i := 0; i < 22; i++ {
		total := i
		for j := cards.ACE; j <= cards.KING; j++ {
			dealerTopCard := cards.CardRank(j)
			decisions[total][dealerTopCard] = softTotalDecision[total][dealerTopCard]
		}
	}

	return decisions
}

// not exported
var _SOFT_TOTAL_DECISIONS [22]map[cards.CardRank]Decision = createSoftTotalDecisions()

func GetSoftTotalDecision(
	playerTotal int, 
	dealerTopCard cards.CardRank,
) Decision {
	var decision Decision = _SOFT_TOTAL_DECISIONS[playerTotal][dealerTopCard]
	return decision
}

File: rust/blackjack/src/strategy/soft.rs
// file src/strategy/soft.rs defines project module "strategy::soft".

use std::collections::HashMap;

use super::Decision;
use super::Decision::Dh;
use super::Decision::Ds;
use super::Decision::Uh;
use super::Decision::Us;
use super::Decision::Usp;
use super::Decision::H;
use super::Decision::NO;
use super::Decision::S;
use super::Decision::SP;

use crate::cards;

// turn off formating for the entire module
#[rustfmt::skip]

const _SOFT_TOTAL_DECISIONS: [[Decision; 14]; 22] = [
    //0   A   2   3   4   5   6   7   8   9  10   J   Q   K
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 1 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 2 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 3 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 4 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 5 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 6 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 7 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 8 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 9 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 10 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 11 x dealer top card
    [NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO, NO],
    // soft hand total 12 x dealer top card
	[NO,  H,  H,  H,  H,  H, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 13 x dealer top card
	[NO,  H,  H,  H,  H, Dh, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 14 x dealer top card
	[NO,  H,  H,  H,  H, Dh, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 15 x dealer top card
	[NO,  H,  H,  H, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 16 x dealer top card
	[NO,  H,  H,  H, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 17 x dealer top card
	[NO,  H,  H, Dh, Dh, Dh, Dh,  H,  H,  H,  H,  H,  H,  H],
    // soft hand total 18 x dealer top card
	[NO,  H, Ds, Ds, Ds, Ds, Ds,  S,  S,  H,  H,  H,  H,  H],
    // soft hand total 19 x dealer top card
	[NO,  S,  S,  S,  S,  S, Ds,  S,  S,  S,  S,  S,  S,  S],
    // soft hand total 20 x dealer top card
	[NO,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S],
    // soft hand total 21 x dealer top card
	[NO,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S,  S],
];

use lazy_static::lazy_static;

lazy_static! {
    static ref SOFT_TOTAL_DECISIONS: Vec<HashMap<cards::CardRank, Decision>> = {
        let mut soft_total_decisions: Vec<HashMap<cards::CardRank, Decision>> = vec![];
        for i in 0..22 {
            let mut decisions_row: HashMap<cards::CardRank, Decision> = HashMap::new();
            for rank in cards::CardRank::iterator() {
                // rank: &cards::CardRank
                let decision: Decision = _SOFT_TOTAL_DECISIONS[i][rank.discriminant()];
                decisions_row.insert(*rank, decision);
                // the trait `Eq` is not implemented for `CardRank`, which is required by `HashMap<_, _, _>: Index<&_>`
                // the trait `Hash` is not implemented for `CardRank`, which is required by `HashMap<_, _, _>: Index<&_>`
            }
            // hard_total_decisions[i] = decisions_row;
            soft_total_decisions.push(decisions_row);
        }
        soft_total_decisions
    };
}

pub fn get_soft_total_decision(
    player_hand_total: usize,
    dealer_top_card: &cards::Card,
) -> Decision {
    let dealer_top_card_rank: cards::CardRank = dealer_top_card.rank;
    let decision: Decision = SOFT_TOTAL_DECISIONS[player_hand_total][&dealer_top_card_rank];
    return decision;
}
